base/compiler/ssair/inlining.jl

```diff
@@ -1349,7 +1291,24 @@ function handle_cases!(ir::IRCode, idx::Int, stmt::Expr, @nospecialize(atype),
         handle_single_case!(ir, idx, stmt, cases[1].item, todo, params)
     elseif length(cases) > 0
         isa(atype, DataType) || return nothing
-        all(case::InliningCase->isa(case.sig, DataType), cases) || return nothing
+        # `ir_inline_unionsplit!` is going to generate `isa` checks corresponding to the
+        # signatures of union-split dispatch candidates in order to simulate the dispatch
+        # semantics, and inline their bodies into each `isa`-conditional block -- and since
+        # we may deal with abstract union-split callsites here, these dispatch candidates
+        # need to be sorted in order of their signature specificity.
+        # Fortunately, ml_matches already sorted them in that way, so we can just process
+        # them in order, as far as we haven't changed their order somewhere up to this point.
+        ncases = length(cases)
+        for i = 1:ncases
+            sigᵢ = cases[i].sig
+            isa(sigᵢ, DataType) || return nothing
+            for j = i+1:ncases
+                sigⱼ = cases[j].sig
+                # since we already bail out from ambiguous case, we can use `morespecific` as
+                # a strict total order of specificity (in a case when they don't have a type intersection)
+                @assert !hasintersect(sigᵢ, sigⱼ) || morespecific(sigᵢ, sigⱼ) "invalid order of dispatch candidate"
```

Comment:
Surprisingly, the test results show that there are cases where these invariants don't hold. Will look into this.

Reply:
It seems like that the assertion fails when we see duplicated signatures (which corresponds to [this line](https://github.com/JuliaLang/julia/blob/529ac5170277cdb654ce03698675ef27d99d1fab/base/compiler/ssair/inlining.jl#L1316) in the previous inlining call analysis, and I carelessly removed in this PR).

I recovered that line with some comments saying when we hit such duplicated signatures, and now this PR seems to work fine with the assertion enabled.